lib/fuzzer/main.zig

```diff
@@ -0,0 +1,508 @@
+
+        return .{
+            .pcs = pcs,
+            .pc_counters = pc_counters,
+            .cache_dir = cache_dir,
+            .coverage_id = hashPCs(pcs),
+        };
+    }
+
+    pub fn start(f: *Fuzzer, test_one: Testee, options: Options) error{OutOfMemory}!void {
+        // we are a well behaved program
+        greet();
+        defer farewell();
+
+        var rng = Prng.init(0);
+        var gpa_impl: std.heap.GeneralPurposeAllocator(.{}) = .{};
+        const a = gpa_impl.allocator();
+
+        var input: ArrayListUnmanaged(u8) = .{};
+        defer input.deinit(a);
+
+        var mutate_scratch: ArrayListUnmanaged(u8) = .{};
+        defer mutate_scratch.deinit(a);
+
+        var all_features: ArrayListUnmanaged(u32) = .{};
+        defer all_features.deinit(a);
+
+        var feature_buffer = try ArrayListUnmanaged(u32).initCapacity(a, InitialFeatureBufferCap);
+        defer feature_buffer.deinit(a);
+
+        // Choose a file name for the coverage based on a hash of the PCs that
+        // will be stored within.
+        const hex_digest = std.fmt.hex(f.coverage_id);
+        const coverage_file_path = "v/" ++ hex_digest ++ "coverage";
+
+        var ip = InputPool.init(f.cache_dir, f.coverage_id);
+        defer ip.deinit();
+
+        // Tracks which PCs have been seen across all runs that do not crash the fuzzer process.
+        // Stored in a memory-mapped file so that it can be shared with other
+        // processes and viewed while the fuzzer is running.
+        const seen_pcs = initCoverageFile(f.cache_dir, coverage_file_path, f.pcs);
+
+        std.log.info("Coverage id is {s}", .{&hex_digest});
+
+        std.log.info(
+            \\Fuzzer booted with a initial corpus of size {}
+            \\F - this input features
+            \\T - new unique features
+        , .{ip.len()});
+        if (ip.len() == 0) {
+            initialCorpusRandom(&ip, &rng);
+        }
+
+        for (options.corpus) |inp| {
+            try runInput(a, test_one, &feature_buffer, f.pc_counters, inp);
+            const features = getLastRunFeatures();
+            if (analyzeFeatures(&ip, features, inp, all_features.items)) {
+                try mergeInput(a, seen_pcs, &all_features, features, f.pc_counters);
+            }
+        }
+
+        // fuzzer main loop
+        while (true) {
+            incrementNumberOfRuns(seen_pcs);
+            input = try selectAndCopyInput(a, &ip, &rng, input);
+            const mutation_seed = rng.next();
+            assert(mutate_scratch.items.len == 0);
+            try mutate.mutate(&input, mutation_seed, &mutate_scratch, a);
+            const input_checksum = checksum(input.items);
+
+            try runInput(a, test_one, &feature_buffer, f.pc_counters, input.items);
+
+            if (input_checksum != checksum(input.items)) {
+                // report the input? it is not very useful since it was written to
+                @panic("user code mutated input!");
+            }
```

Comment:
The cast can be casted away.

I have a plan of doing in-place mutations and avoiding memcpys of the input string and in this future scenario, user mutating the input would break the fuzzer.